src/rust/demikernel/bindings.rs

```diff
@@ -793,15 +789,63 @@ pub extern "C" fn demi_setsockopt(
     optval: *const c_void,
     optlen: Socklen,
 ) -> c_int {
-    // TODO: Implement this system call.
-    libc::ENOSYS
+    trace!("demi_setsockopt()");
+
+    // Check inputs.
+    if level != SOL_SOCKET {
+        let cause: String = format!("Only options at the socket level are supported");
+        error!("demi_setsockopt(): {}", cause);
+        return libc::EINVAL;
+    }
+
+    let opt: SocketOption = match optname {
+        SO_LINGER => {
+            // Check for invalid storage locations.
+            if optval.is_null() {
+                warn!("demi_setsockopt() linger value is a null pointer");
+                return libc::EINVAL;
+            }
+
+            if optlen as usize != mem::size_of::<Linger>() {
+                warn!("demi_setsockopt() linger len is incorrect");
+                return libc::EINVAL;
+            }
+
+            let linger: Linger = unsafe { *(optval as *const Linger) };
+            match linger.l_onoff {
+                0 => SocketOption::SO_LINGER(None),
+                _ => SocketOption::SO_LINGER(Some(Duration::from_secs(linger.l_linger as u64))),
+            }
```

Comment:
Going back to my previous comment. How does one set linger to 0, if `Non` means default value?

Reply:
you set it to Some(Duration::ZERO)